common/utf8validator/validate.go

```diff
@@ -0,0 +1,247 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package utf8validator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync/atomic"
+	"unicode/utf8"
+
+	"go.temporal.io/api/serviceerror"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+)
+
+type (
+	// A Validator validates utf-8 in string fields of proto messages. The proto library
+	// normally does this automatically, but we disable it for now as part of migration from
+	// gogoproto, which did not validate. Eventually we'll remove this and enable it normally.
+	Validator struct {
+		logger  log.Logger
+		metrics metrics.Handler
+
+		checkBySource [sourceCount]dynamicconfig.FloatPropertyFn
+		failBySource  [sourceCount]dynamicconfig.BoolPropertyFn
+	}
+
+	MessageSource int
+
+	validation struct {
+		badPath string
+	}
+)
+
+const (
+	SourceRPCRequest MessageSource = iota
+	SourceRPCResponse
+	SourcePersistence
+	sourceCount
+)
+
+var (
+	ErrInvalidUTF8          = errors.New("invalid utf-8 in string field")
+	ErrInvalidMessageSource = errors.New("invalid message source")
+
+	sourceNames = [sourceCount]string{
+		"rpc-request",
+		"rpc-response",
+		"persistence",
+	}
+
+	// globalValidator holds one instance of Validator for use where we want to validate but
+	// have no context to hold a reference to a Validator. If multiple services are running in
+	// the same process it will be the last one constructed, but it doesn't matter much since
+	// they should all behave the same (just with different log/metrics tags).
+	globalValidator atomic.Value
+)
+
+func newValidator(
+	logger log.Logger,
+	metrics metrics.Handler,
+	checkRPCRequest dynamicconfig.FloatPropertyFn,
+	checkRPCResponse dynamicconfig.FloatPropertyFn,
+	checkPersistence dynamicconfig.FloatPropertyFn,
+	failRPCRequest dynamicconfig.BoolPropertyFn,
+	failRPCResponse dynamicconfig.BoolPropertyFn,
+	failPersistence dynamicconfig.BoolPropertyFn,
+) *Validator {
+	v := &Validator{
+		logger:  logger,
+		metrics: metrics,
+		checkBySource: [sourceCount]dynamicconfig.FloatPropertyFn{
+			checkRPCRequest, checkRPCResponse, checkPersistence,
+		},
+		failBySource: [sourceCount]dynamicconfig.BoolPropertyFn{
+			failRPCRequest, failRPCResponse, failPersistence,
+		},
+	}
+	globalValidator.Store(v)
+	return v
+}
+
+// ValidateUsingGlobalValidator maybe validates one proto message, depending on dynamic config
+// and random sampling. If validation fails, this may still return no error, depending on
+// dynamic config. This deliberately does not return a serviceerror because the decision of
+// which serviceerror to use is contextual. The caller should wrap it in the appropriate
+// serviceerror.
+func ValidateUsingGlobalValidator(
```

Comment:
I'd prefer we name it `Validate` rather than `ValidateUsingGlobalValidator`. It's clear it's using some kind of global state as it is neither a message hanging off of the validator class nor takes one as a parameter.

This would be in line with how the JSON and other packages handle their naming (it's json.Marshal, not json.MarshalWithGlobal), thought `encoding/base64` defies this by forcing you to use `base64.StdEncoding.Foo` rather than some top-level base64.Foo method